hw/ip/otbn/dv/rig/rig/gens/bad_at_end.py

```diff
@@ -0,0 +1,123 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import random
+from typing import List, Optional, Tuple
+
+from shared.insn_yaml import InsnsFile
+
+from .branch import Branch
+from .jump import Jump
+from .loop import Loop
+from ..config import Config
+from ..program import ProgInsn, Program
+from ..model import Model
+from ..snippet_gen import GenCont, GenRet
+
+
+class BadAtEnd(Loop):
+    '''A snippet generator that generates a loop/branch/jump at end of a loop
+
+    This works by overriding the _gen_tail method of the Loop generator and
+    replacing the last instruction that we would have generated with a loop,
+    branch or jump instruction.
+
+    '''
+
+    ends_program = True
+
+    def __init__(self, cfg: Config, insns_file: InsnsFile) -> None:
+        super().__init__(cfg, insns_file)
+        self.branch_gen = Branch(cfg, insns_file)
+        self.jump_gen = Jump(cfg, insns_file)
+
+    def pick_weight(self,
+                    model: Model,
+                    program: Program) -> float:
+        # Only try this if we've got a reasonable amount of room.
+        room = min(model.fuel, program.space)
+        assert 0 < room
+        return (1.0 if room > 50 else 0.0)
+
+    def _gen_tail(self,
+                  num_insns: int,
+                  model: Model,
+                  program: Program) -> Optional[Tuple[List[ProgInsn], Model]]:
+        assert num_insns > 0
+        old_model = model.copy()
+        ret = super()._gen_tail(num_insns, model, program)
+        if ret is None:
+            return None
+
+        insns, model = ret
+        assert len(insns) == num_insns
+
+        # We need a model to represent the state just before the last
+        # instruction. Of course, we can't "unwind" the state update that we
+        # just did. However, a merge between the initial state and the end
+        # state gives a sound (if very imprecise!) approximation.
+        #
+        # Since the Model.merge() method expects the two models to have the
+        # same PC (which is the usual situation when implementing phi nodes),
+        # we also have to teleport old_model into the right place!
+        old_model.pc = model.pc
+        model.merge(old_model)
+
+        # Pick a new instruction to replace the last one we generated. Start by
+        # shuffling the numbers from 0..2, which will represent Branch, Loop,
+        # Jump respectively. The idea is that we want to try them in an
+        # arbitrary order, but don't want to fail completely if one of the
+        # other generators might have worked.
+        types = list(range(2))
```

Comment:
Should be `range(3)` most likely?

Reply:
Ah, that explains why some of the coverage points weren't being hit much :-)